packages/runtime/container-runtime/src/garbageCollection.ts

```diff
@@ -69,6 +77,67 @@ export interface IGarbageCollector {
     ): Promise<IGCStats>;
     /** Called when the latest summary of the system has been refreshed. */
     latestSummaryStateRefreshed(result: RefreshSummaryResult, readAndParseBlob: ReadAndParseBlob): Promise<void>;
+    /** Called when a node is changed. Used to detect and log when an inactive node is changed. */
+    nodeChanged(id: string): void;
+}
+
+/** The garbage collection data of each node in the reference graph. */
+interface IGCNodeData {
+    /** The set of routes to other nodes in the graph. */
+    outboundRoutes: string[];
+    /** If the node is unreferenced, the timestamp of when it was marked unreferenced. */
+    unreferencedTimestampMs?: number;
+}
+
+/**
+ * The garbage collection state of the reference graph. It contains a list of all the nodes in the graph and their
+ * GC data.
+ */
+interface IGCState {
+    gcNodes: { [ id: string ]: IGCNodeData };
+}
+
+/**
+ * Helper class that tracks the state of an unreferenced node such as the time it was unreferenced. It also sets
+ * the node's state to inactive if it remains unreferenced for a given amount of time (inactiveTimeoutMs).
+ */
+class UnreferencedStateTracker {
+    private inactive: boolean = false;
+    // Keeps track of all inactive events that are logged. This is used to limit the log generation for each event to 1
+    // so that it is not noisy.
+    private readonly inactiveEventsLogged: Set<string> = new Set();
+
+    constructor(
+        public readonly unrefencedTimestampMs: number,
+        inactiveTimeoutMs: number,
+    ) {
+        // If the timeout has already expired, the node should become inactive immediately. Otherwise, start a timer of
+        // inactiveTimeoutMs after which the node will become inactive.
+        if (inactiveTimeoutMs <= 0) {
+            this.inactive = true;
+        } else {
+            setTimeout(() => { this.inactive = true; }, inactiveTimeoutMs);
```

Comment:
What is the way for this timer to get cancelled if node is used again?

Reply:
I didn't add cancelation for the timer because the node itself is deleted when the node is referenced. The timer will still keep on running. Maybe I should cancel it explicitly.

Reply:
Added explicit clear of timers.